src/main/java/Problem1/ValidParentheses.java

```diff
@@ -4,8 +4,92 @@
 
     // Do not change signature (function name, parameters, return type)
     public static boolean isValid(String str) {
-        // homework
-        // !!! must use ArrayStack or LinkedListStack from problem 1
-        return false; // place holder
+        if(str==null){
+            return true;
+        }
+        if(str.length()==0){
+            return true;
+        }
+        LinkedListStack<String> parens = new LinkedListStack<String>();
+        LinkedListStack<String> brackets = new LinkedListStack<String>();
+        LinkedListStack<String> curly = new LinkedListStack<String>();
+        LinkedListStack<String> other = new LinkedListStack<String>();
+        for(String s : str.split("")){
+            switch (s) {
+                case "(":
+                case ")":
+                    parens.push(s);
+                    break;
+                case "{":
+                case "}":
+                    curly.push(s);
+                    break;
+                case "[":
+                case "]":
+                    brackets.push(s);
+                    break;
+                default:
+                    other.push(s);
+                    break;
+            }
+        }
+        if(other.size()>0){
+            return false;
+        }
+
+        if(!pairChecker(parens, ")")){
+            return false;
+        }
+
+
+        if(!pairChecker(curly, "}")){
+            return false;
+        }
+
+        if(!pairChecker(brackets, "]")){
+            return false;
+        }
+
+        return true;
+    }
+
+    private static boolean pairChecker(LinkedListStack<String> stack, String close){
+        if(stack.size()==0){
+            return true;
+        }
+        String s = stack.pop();
+        //if we have a closing character, check for any open characters
+        if(s.equals(close)){
+            return recur(stack, close, 1);
+        }
+        //only other option is that it equals a close
+        else
+            return false;
+    }
+    private static boolean recur(LinkedListStack<String> stack, String close, int numClose){
+        //if we just saw an open but nothing is left on the stack, its invalid
+        if(stack.size()==0){
+            return false;
+        }
+        String s = stack.pop();
+        //if we have a closing character, check for any open characters
+        if(s.equals(close)){
+            return recur(stack, close, numClose+1);
```

Comment:
very interesting implementation. 

I don't agree though recursion is a very efficient method for this purpose. recursion has very high overhead cost and here we could just implement it using a loop.